src/base_alloc/base_alloc_global.c

```diff
@@ -5,40 +5,94 @@
  * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 */
 
-#include <assert.h>
+/* A MT-safe base allocator */
 
+#include "base_alloc_global.h"
 #include "base_alloc.h"
+#include "base_alloc_internal.h"
+#include "utils_math.h"
 
-#define SIZE_BA_POOL_CHUNK 128
+#include <stdio.h>
 
-// global base allocator used by all providers and pools
-static umf_ba_pool_t *BA_pool = NULL;
+// allocation classes need to be powers of 2
+#define ALLOCATION_CLASSES                                                     \
+    { 16, 32, 64, 128 }
+#define NUM_ALLOCATION_CLASSES 4
 
-int umf_ba_create_global(void) {
-    assert(BA_pool == NULL);
+struct base_alloc_t {
+    size_t ac_sizes[NUM_ALLOCATION_CLASSES];
+    umf_ba_pool_t *ac[NUM_ALLOCATION_CLASSES];
+};
 
-    BA_pool = umf_ba_create(SIZE_BA_POOL_CHUNK);
-    if (!BA_pool) {
-        return -1;
+static struct base_alloc_t BASE_ALLOC = {.ac_sizes = ALLOCATION_CLASSES};
+static size_t SMALLEST_AC_SIZE_LOG2 = 0;
+
+void umf_ba_create_global(void) {
+    for (int i = 0; i < NUM_ALLOCATION_CLASSES; i++) {
+        // allocation classes need to be powers of 2
+        assert(0 == (BASE_ALLOC.ac_sizes[i] & (BASE_ALLOC.ac_sizes[i] - 1)));
+        BASE_ALLOC.ac[i] = umf_ba_create(BASE_ALLOC.ac_sizes[i]);
+        if (!BASE_ALLOC.ac[i]) {
+            fprintf(stderr,
+                    "Cannot create base alloc allocation class for size: %zu\n",
+                    BASE_ALLOC.ac_sizes[i]);
+        }
     }
 
-    return 0;
+    size_t smallestSize = BASE_ALLOC.ac_sizes[0];
+    SMALLEST_AC_SIZE_LOG2 = log2Utils(smallestSize);
 }
 
 void umf_ba_destroy_global(void) {
-    assert(BA_pool);
-    umf_ba_destroy(BA_pool);
-    BA_pool = NULL;
+    for (int i = 0; i < NUM_ALLOCATION_CLASSES; i++) {
+        if (BASE_ALLOC.ac[i]) {
```

Comment:
we call ba_create_global in the lib constructor - what would be the scenario where alloc class creation would fail? For me, allowing allocation using ba_os_alloc here could hide some problems even if everything would work, If we want to revert to a different allocator, we can always do it on the caller side

Reply:
Whenever `umf_ba_create` can fail. I think the general strategy should be to not assert things that we can easily recover from. I get that it makes it easier to catch unexpected behaviors, but on the other hand, there might be corner cases we might miss, and you'll get a crash in the Release version. For things like this, I'd say it's better to just log it. Once we have a logging framework we can even have some kind of switch that can enable asserts for all LOG(ERROR) in debug (so that we can run tests with it) but will not fail in Release.

Reply:
Yes, we already log it (line 37). I've added extra emphasis that this is an error.